Move to new APEX.amp API

justusschock opened the pull request:
Closes #71 

Re-Enables current (unified and more stable) AMP API.

Apex must still be installed manually and following [this PR](https://github.com/NVIDIA/apex/pull/252) it's mixed precision training currently only works on GPU.

EDIT: Switched base branch to not intermingle this with trainer refactoring

justusschock committed:
Move to new APEX.amp API

mibaumgartner commented on delira/training/pytorch_trainer.py:
```diff
@@ -52,9 +52,14 @@ def __init__(self,
-                     mixed_precision_kwargs={"enable_caching": True,
-                                             "verbose": False,
-                                             "allow_banned": False},
+                     mixed_precision_kwargs={"opt_level": "o0",
```
Not sure if our default value should be `o0` because we have an additional `mixed_precision` argument. If I set `mixed_precision` to true, I would assume that some kind of mixed precision training is used and not full fp32 training.
I think we should either set the default value to `o1` (should be equivalent to the old implementation?) or rename `mixed_precision` to something more appropriate like `use_amp`.

I think, I would prefer the different default value.

justusschock replied:
Changed by https://github.com/justusschock/delira/pull/97/commits/55d9ca898108e9acd99e1b651de336410dfeeec9

mibaumgartner commented on delira/models/model_utils.py:
```diff
@@ -0,0 +1,26 @@
+
+
+@contextlib.contextmanager
+def scale_loss(loss,
```
Maybe add `**kwargs` if the apex team decides to add some additional arguments in the future.

justusschock replied:
Added by https://github.com/justusschock/delira/pull/97/commits/ea687b3411b5cd61ea49c455da263c40fd1ca3d8

mibaumgartner reviewed (approved):
Left two comments which are a personal preference. The rest looks good to me 👍

justusschock committed:
Change default mixed precision

Change default kwargs for mixed precision (only used if  `mixed_precision` is set to `True`)

justusschock committed:
Add kwargs to decorator function
